### PR TITLE
Use a tree data structure to build template digests

### PR DESCRIPTION
**Pull request by tenderlove**
This PR changes the template digest code to build a dependency tree data structure before calculating the fingerprint and child finger prints.  I carefully made the tree produce exactly the same digests that the previous code produced (if you read the commits you'll see that I tested them side by side).
## Pros
- Slightly shorter implementation
- Easier to extend (we can just add new tree node types)
- Less memory consumption (I'll talk about this later)
- Easier to debug (I'll talk about that later too)
## Cons
- Possibly slower key calculation (I'll also talk about this later)
### Memory consumption / key calculation speed

The previous implementation would cache a digest for all nodes in any particular tree.  This would speed up digest calculation for templates that have common subtrees.  The downside is that any nodes that are not in common waste memory and once the app is warmed up, those keys are also wasting memory (since we only ever look up a digest by it's root node).  IOW, [our in-memory cache](https://github.com/rails/rails/blob/5971fddb7997231f3431c26ba17471ba52704852/actionview/lib/action_view/digestor.rb#L8) is always wasting memory.

The implementation I'm proposing here only caches the digest for the top most node (which is the node callers of `digest` were asking about).

Key calculation time should be amortized over the life of the app, so I don't think slowing down some calculations should be a big deal, but I'm happy to change if this turns out to be an issue.  Fortunately all we have to do to support that is pass the cache object down as we walk the tree.
### Strange things I found

If you have a wildcard dependency, like this:

```
<% # Template Dependency: users/* %>
```

_Everything_ in `users/` will be considered a "dependency", even files that are not templates.  Here is a visualization of the dependency tree for a test app:

tree

These files that are not templates will be considered as "missing" and the parent template will end up [getting an empty string returned for that template](https://github.com/rails/rails/blob/5971fddb7997231f3431c26ba17471ba52704852/actionview/lib/action_view/digestor.rb#L71) and [will digest a string like "--"](https://github.com/rails/rails/blob/5971fddb7997231f3431c26ba17471ba52704852/actionview/lib/action_view/digestor.rb#L114-L120).  The point being that these so-called "missing dependencies" can impact the digest of the parent template even though we don't care about them.  If I delete the file "users/show" from the tree in the picture, it would impact the digest for "users/new" even though "users/new" doesn't care whats inside "users/show" at all!  This test demonstrates the issue:

``` diff
+  def test_non_partials_do_not_impact_wildcard_dependencies
+    old_caching, ActionView::Resolver.caching = ActionView::Resolver.caching, false
+    before = digest "events/index"
+    add_template 'events/show'
+
+    ActionView::Digestor.cache.clear
+    assert_equal before, digest("events/index")
+  ensure
+    ActionView::Resolver.caching = old_caching
+  end
```

This test fails with the current implementation.  I would like to remove this "feature" and fix the above test.

If we're OK with this PR I'd also like to add a rake task that outputs the dependency tree as a dot file so we can get a visual representation of the deps.

### CODE REVIEW

File: actionview/lib/action_view/digestor.rb [hunk 1]
```diff
@@ -25,108 +25,100 @@ class << self
       def digest(name:, finder:, **options)
         options.assert_valid_keys(:dependencies, :partial)
```
Comment (rafaelfranca) [inline actionview/lib/action_view/digestor.rb]: Using the kwargs for these since `options` can only have these two keys?

**Comment by rafaelfranca** [comment]
The implementation is easier and performance can be amortized as you said so for me it is :shipit:

File: actionview/lib/action_view/digestor.rb (hunk 1 above)
Comment (tenderlove) [inline actionview/lib/action_view/digestor.rb]: Derp, ya, I'll change to kwargs.

**Comment by tenderlove** [comment]
I'm slightly concerned about changing the caching semantics before 5-0-stable is cut.  Our app doesn't use this yet (but I'm really excited to start) so I can't really get a good idea of the "real impact".

Another strange thing I noticed is that the original code will [disable cache on the template finder](https://github.com/rails/rails/blob/310e4418f4db899dcd07ff24f530ef8862c65f8b/actionview/lib/action_view/digestor.rb#L107).  I *think* the original intent was so that template changes will get picked up (it was originally introduced in 4725d588), but this code is not effective because once the top level node is cached, the templates dependencies are never traversed again.  If we're really concerned about cache key generation speed, we could stop disabling the template lookup cache, then we're only paying to md5 the template contents (which should be pretty darn fast).

/cc @jeremy

**Comment by jeremy** [comment]
Could you mimic existing caching behavior, to start, then chop it after 5-0-stable?

**Comment by tenderlove** [comment]
@jeremy ya, I could do that. Any idea when we're branching tho?

**Comment by thedarkone** [comment]
Since we're no longer using `@@cache` to track recursion (using stack local `seen={}` instead), I think we can kill off `@@digest_mutex` altogether.

Something like this is now completely fine:
```ruby
# VERSION 1.
# let multiple threads potentially do redundant work
@@cache[cache_key] ||= tree(name, finder, dependencies).digest
# VERSION 2.
# or rely on TS::Map's concurrency primitive (no redundant work, allowing for possible concurrency on jruby/rbx for different cache_keys)
@@cache.compute_if_absent(cache_key) { tree(name, finder, dependencies).digest }
```
------

But, but... you promised to remove the recursion :crying_cat_face:!

**Comment by tenderlove** [comment]
@thedarkone I mean recursion for *that* method! ;)  My love for functional programming won't allow me to remove all the recursion.

**Pull request merged**